src/components/app-state.js

```diff
@@ -0,0 +1,60 @@
+import React, { createContext, useReducer } from "react";
+import { merge, getTimeString } from "./util";
+
+const initialState = {
+  user: null,
+  entities: {},
+  lists: {},
+  reqs: {}
+};
+
+const reducer = (state, action) => {
+  state = { ...state };
+
+  for (let key in action) {
+    if (key === "entities") {
+      state.entities = merge(state.entities, action.entities, 3);
+    } else if (key === "lists") {
+      state[key] = merge(state[key], action[key], 2);
+    } else if (key === "reqs") {
+      state[key] = merge(state[key], action[key]);
+    } else {
+      state[key] = action[key];
+    }
+  }
+
+  if (process.env.NODE_ENV !== "production" && typeof window !== "undefined") {
+    console.group(getTimeString(new Date()));
+    console.log(state);
+    console.groupEnd();
+  }
+
+  return state;
+};
+
+export const State = createContext();
+export const Dispatch = createContext();
+
+export default ({ initialState, children }) => {
+  const [state, dispatch] = useReducer(reducer, initialState);
+
+  return (
+    <State.Provider value={state}>
+      <Dispatch.Provider value={dispatch}>{children}</Dispatch.Provider>
+    </State.Provider>
```

Comment:
It is separated for the performance concern.

Consider this component:

```js
function DoIt() {
  const dispatch = useContext(Dispatch);
  return <button onClick={() => dispatch({ done: true })}>Do It</button>
}
```

A update in `State` will not cause `<DoIt />` to render, since `dispatch` is static in the whole life cycle of the application.

But if `<DoIt />` relies on a single context which contains both State and Dispatch, it will have to perform many unnecesary `rerender`s.